File: src/TradingSystem.Web/Services/SubscriptionApiService.cs
using System.Net.Http.Json;
using TradingSystem.Web.Models;

namespace TradingSystem.Web.Services;

public interface ISubscriptionApiService
{
    Task<SubscriptionInfo?> GetStatusAsync();
    Task<List<PlanTier>> GetPlansAsync();
    Task<UpgradeResponse?> UpgradeAsync(UpgradeRequest request);
    Task<UpgradeResponse?> CancelAsync();
    Task<bool> CheckFeatureAccessAsync(string feature);
}

public class SubscriptionApiService : ISubscriptionApiService
{
    private readonly HttpClient _httpClient;

    public SubscriptionApiService(HttpClient httpClient) => _httpClient = httpClient;

    public async Task<SubscriptionInfo?> GetStatusAsync()
    {
        try
        {
            var resp = await _httpClient.GetFromJsonAsync<ApiResponse<SubscriptionInfo>>("/subscription/status");
            return resp?.Data;
        }
        catch { return null; }
    }

    public async Task<List<PlanTier>> GetPlansAsync()
    {
        try
        {
            var resp = await _httpClient.GetFromJsonAsync<ApiResponse<List<PlanTier>>>("/subscription/plans");
            return resp?.Data ?? new();
        }
        catch { return new(); }
    }

    public async Task<UpgradeResponse?> UpgradeAsync(UpgradeRequest request)
    {
        try
        {
            var resp = await _httpClient.PostAsJsonAsync("/subscription/upgrade", request);
            return await resp.Content.ReadFromJsonAsync<UpgradeResponse>();
        }
        catch { return null; }
    }

    public async Task<UpgradeResponse?> CancelAsync()
    {
        try
        {
            var resp = await _httpClient.PostAsync("/subscription/cancel", null);
            return await resp.Content.ReadFromJsonAsync<UpgradeResponse>();
        }
        catch { return null; }
    }

    public async Task<bool> CheckFeatureAccessAsync(string feature)
    {
        try
        {
            var resp = await _httpClient.GetFromJsonAsync<dynamic>($"/subscription/check-feature/{feature}");
            return true;
        }
        catch { return false; }
    }
}

File: src/TradingSystem.Angular/src/app/services/subscription-api.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { ApiResponse, SubscriptionInfo, PlanTier, UpgradeRequest, UpgradeResponse } from '../models/models';

@Injectable({ providedIn: 'root' })
export class SubscriptionApiService {
  constructor(private http: HttpClient) {}

  getStatus(): Observable<SubscriptionInfo | null> {
    return this.http.get<ApiResponse<SubscriptionInfo>>('/subscription/status').pipe(
      map(r => r.data || null),
      catchError(() => of(null))
    );
  }

  getPlans(): Observable<PlanTier[]> {
    return this.http.get<ApiResponse<PlanTier[]>>('/subscription/plans').pipe(
      map(r => r.data || []),
      catchError(() => of([]))
    );
  }

  upgrade(request: UpgradeRequest): Observable<UpgradeResponse | null> {
    return this.http.post<UpgradeResponse>('/subscription/upgrade', request).pipe(
      catchError(() => of(null))
    );
  }

  cancel(): Observable<UpgradeResponse | null> {
    return this.http.post<UpgradeResponse>('/subscription/cancel', null).pipe(
      catchError(() => of(null))
    );
  }

  checkFeatureAccess(feature: string): Observable<boolean> {
    return this.http.get(`/subscription/check-feature/${feature}`).pipe(
      map(() => true),
      catchError(() => of(false))
    );
  }
}
